### ISSUE TITLE
Example where pylint does not detect uninitialized variable

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#4045]
```python
def test():
    try:
        with open('xxxxx', 'rt') as config_file:
            old_settings: str = config_file.read()
    except FileNotFoundError:
        print(old_settings)
```
Pylint should report that `old_settings` might not be initialized. Since initialization is the last operation in try...except block, any exception prevents initialization.

**Comment by hippo91** [comment]
@socketpair thanks for the report. It makes sense.

**Comment by jacobtylerwalls** [comment]
On main, this is only reproducible because of the context manager, which fails the following check (the context manager gets between the assignment and the TryExcept, so the TryExcept is actually _two_ parents up):

https://github.com/PyCQA/pylint/blob/52576b7552b68593a3ea28e20cd3751ccdc106d7/pylint/checkers/variables.py#L686

**Comment by DanielNoord** [comment]
Might be fixable with `checkers.utils.get_node_first_ancestor_of_type`?

**Comment by jacobtylerwalls** [comment]
You can assign this to me, I'll give it a spin and report back.

**Comment by markmark206** [comment]
And what about this case (should I open a separate issue?): 

```
$ cat use_before_assignment.py
def compute() -> int:
    raise ValueError("bad chickens")

def count_chickens():
    try:
        chicken_count = compute()
    except ValueError:
        print(chicken_count)

count_chickens()
```
```
$ pylint -d missing-module-docstring,missing-function-docstring use_before_assignment.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
```
$ pylint --version
pylint 2.12.2
astroid 2.9.3
Python 3.9.9 (v3.9.9:ccb0e6a345, Nov 15 2021, 13:06:05)
[Clang 13.0.0 (clang-1300.0.29.3)]
```
```
$ python use_before_assignment.py
Traceback (most recent call last):
  File "/Users/markgenba/src/genba/use_before_assignment.py", line 6, in count_chickens
    chicken_count = compute()
  File "/Users/markgenba/src/genba/use_before_assignment.py", line 2, in compute
    raise ValueError("bad chickens")
ValueError: bad chickens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/markgenba/src/genba/use_before_assignment.py", line 10, in <module>
    count_chickens()
  File "/Users/markgenba/src/genba/use_before_assignment.py", line 8, in count_chickens
    print(chicken_count)
UnboundLocalError: local variable 'chicken_count' referenced before assignment
```

**Comment by DanielNoord** [comment]
@markmark206 This actually does emit a warning on the to-be-released `2.13` version thanks to the work of @jacobtylerwalls  and others!

**Comment by markmark206** [comment]
@DanielNoord this is very exciting, thank you!! 

I opened a separate issue before i saw this, I will go ahead and close (unless you get there before me).

Thank you again!!
